heat/core/communication.py

```diff
@@ -463,6 +464,161 @@ def Iallgatherv(self, sendbuf, recvbuf, send_axis=0, recv_axis=None):
+
+        # unpack the receive buffer
+        if isinstance(recvbuf, tuple):
+            recvbuf, recv_counts, recv_displs = recvbuf
+        if isinstance(recvbuf, dndarray.DNDarray):
+            recvbuf = recvbuf._DNDarray__array
+        if not isinstance(recvbuf, torch.Tensor) and send_axis != 0:
+            raise TypeError('recvbuf of type {} does not support send_axis != 0'.format(type(recvbuf)))
+
+        # keep a reference to the original buffer object
+        original_recvbuf = recvbuf
+
+        # Send_axis-Permutation: [recv_axis, send_axis, rest ...]
+        axis_permutation = list(range(recvbuf.ndimension()))
+        if(send_axis == 0):
+            if(recv_axis == 1 ):
+                axis_permutation[send_axis], axis_permutation[recv_axis] = recv_axis, send_axis
+            else:
+                axis_permutation[1], axis_permutation[send_axis] = send_axis, 1
+                axis_permutation[recv_axis] = axis_permutation[0]
+                axis_permutation[0] = recv_axis
+
+        else:
+            axis_permutation[0], axis_permutation[recv_axis] = recv_axis, 0
+            axis_permutation[send_axis] = axis_permutation[1]
+            axis_permutation[1] = send_axis
+
+        sendbuf = sendbuf.permute(*axis_permutation)
+        recvbuf = recvbuf.permute(*axis_permutation)
+
+        # prepare buffer objects
+        nproc = self.size
+        #if sendbuf is not MPI.IN_PLACE: --> Can we allow IN_Place opperation? (Buffer config might be wrong
+        mpi_sendbuf = self.alltoall_sendbuffer(sendbuf, nproc, send_counts, send_displs)
+        #else:
+        #   mpi_sendbuf = sendbuf
+
+        #if recvbuf is not MPI.IN_PLACE:
+        mpi_recvbuf = self.alltoall_recvbuffer(recvbuf, nproc, recv_counts, recv_displs)
+        #else:
+        #    mpi_recvbuf = recvbuf
+
+        exit_code = self.handle.Alltoallw(mpi_sendbuf, mpi_recvbuf, **kwargs)
+        original_recvbuf.set_(recvbuf.storage(), recvbuf.storage_offset(), original_recvbuf.shape, original_recvbuf.stride())
+
+        return exit_code
+
+
+    def Alltoall(self, sendbuf, recvbuf, axis, recv_axis):
+        raise NotImplementedError(
+            'AllToAll for same axes not supported. Please choose send_axis and recv_axis to be different.')
+        if(((axis == 0) & (recv_axis == 1)) | ((axis == 1) & (recv_axis == 0))):
+            return self.__scatter_like(self.handle.Alltoall, sendbuf, recvbuf, axis, recv_axis, send_factor=self.size, recv_factor=self.size)
+        else:
+            return self.__alltoall_w(sendbuf, recvbuf, axis, recv_axis)
+    Alltoall.__doc__ = MPI.Comm.Alltoall.__doc__
+
+    def Alltoallv(self, sendbuf, recvbuf, axis, recv_axis):
+        raise NotImplementedError(
+            'AllToAll for same axes not supported. Please choose send_axis and recv_axis to be different.')
+        if(((axis == 0) & (recv_axis == 1)) | ((axis == 1) & (recv_axis == 0))):
```

Comment:
This is most likely not what you want to do here. You are looking for the logical operators 'and' and 'or' here instead of the bitwise & and |

Reply:
this is my C++ speaking :laughing: